### PR TITLE
`build/bin/sage-spkg`: Add support for installing script packages

### LINKED ISSUE #29386: Install script and pip packages via sage-spkg
<div id="comment:0"></div>

Follow-up from #29287.

Script packages are installed by special rules in `build/make/Makefile`.

This ticket installs them by `sage-spkg` instead. This enables script packages to participate in the staged installs (`SAGE_DESTDIR`) and uninstallation mechanisms.

This is useful for #29387 to clean up #27870/#29003/#29082, which introduced files in `SAGE_LOCAL` that do not belong to any package and are never uninstalled.

We also make the decision whether `spkg-install` is generated from a template `spkg-install.in` or provided as is in the source tree orthogonal to the distinction `source=script` vs. `source=pip`.
We simply generate `spkg-install` if `spkg-install.in` is present.

Ideally, when done, changing a package's source between script and normal and pip will no longer require running `bootstrap`.

CC:  @dimpase @jhpalmieri @embray @orlitzky

Component: **build**

Work Issues: **Move "dummy script package" code to sage-spkg**

Author: **Matthias Koeppe**

Branch/Commit: **[u/mkoeppe/install_script_packages_via_sage_spkg](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/install_script_packages_via_sage_spkg) @ [`7a25d1d`](https://github.com/sagemath/sagetrac-mirror/commit/7a25d1d0b74cf7fe7b29b0169ba2539b323ae04a)**

_Issue created by migration from https://trac.sagemath.org/ticket/29386_

### PR DESCRIPTION
**Pull request by mkoeppe**
Split out and updated from #29386. 

The section of `build/make/Makefile.in` that deals with script packages can be consolidated with the section on normal packages later, in a follow-up PR after #36740, #36738.

We switch the `spkg-install` scripts of the `sage*` script packages to the templated versions (`spkg-install.in`), which is a simplification. `spkg-check` remains as is because we still invoke it directly (changing this will need #36738).

Resolves #29386.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- Depends on #36737 (for testing)

### CODE REVIEW

**Comment by orlitzky** [comment]
I get two errors under "normal" circumstances that don't hurt anything:

```
$ sage-spkg sage_conf
...
Package 'sage_conf' is currently not installed
No legacy uninstaller found for 'sage_conf'; nothing to do
Installing sage_conf-10.2.rc4
Looking in links: /home/mjo/src/sage.git/local/var/lib/sage/wheels
Processing /home/mjo/src/sage.git/pkgs/sage-conf
  Running command pip subprocess to install build dependencies
  Looking in links: /home/mjo/src/sage.git/local/var/lib/sage/wheels
  WARNING: Location '/home/mjo/src/sage.git/local/var/lib/sage/wheels' is ignored: it is either a non-existing path or lacks a specific scheme.
  ERROR: Could not find a version that satisfies the requirement setuptools (from versions: none)
  ERROR: No matching distribution found for setuptools
  error: subprocess-exited-with-error
  
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /home/mjo/src/sage.git/local/var/lib/sage/venv-python3.11/bin/python3 /usr/lib/python3.11/site-packages/pip/__pip-runner__.py install --ignore-installed --no-user --prefix /tmp/pip-build-env-eto9101s/overlay --no-warn-script-location --no-binary :none: --only-binary :none: --no-index --find-links /home/mjo/src/sage.git/local/var/lib/sage/wheels -- setuptools wheel
  cwd: [inherit]
  Installing build dependencies ... error
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
Warning: building with "python3 -m pip wheel --wheel-dir=dist --verbose --no-deps --no-index --isolated --ignore-requires-python --find-links=/home/mjo/src/sage.git/local/var/lib/sage/wheels" failed.
Retrying with "python3 -m pip wheel --wheel-dir=dist --verbose --no-deps --no-index --isolated --ignore-requires-python --no-build-isolation --no-binary :all:".
```

The retry works, but the error is in bright red and always happens AFAICT so it would be better if, well, it didn't. It may have something to do with `--enable-system-site-packages` since I'm using almost everything (and in particular *setuptools* and *wheel*) from the system.

The other error can probably be fixed with an `-f` test?

```
$ sage-spkg-uninstall sage_conf
Uninstalling existing 'sage_conf'
Running pip-uninstall script for 'sage_conf'
ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/home/mjo/src/sage.git/local/var/lib/sage/scripts/sage_conf/spkg-requirements.txt'
Warning: pip exited with status 0
Removing stamp file '/home/mjo/src/sage.git/local/var/lib/sage/installed/sage_conf-10.2.rc4'
```

**Comment by mkoeppe** [comment]
> It may have something to do with `--enable-system-site-packages` since I'm using almost everything (and in particular _setuptools_ and _wheel_) from the system.

The first try uses build isolation, so it needs the wheel file of setuptools.

**Comment by mkoeppe** [comment]
... so perhaps when using `--enable-system-site-packages`, you would want to disable build isolation?

**Comment by mkoeppe** [comment]
> The other error can probably be fixed with an `-f` test?
> 
> ```
> $ sage-spkg-uninstall sage_conf
> Uninstalling existing 'sage_conf'
> Running pip-uninstall script for 'sage_conf'
> ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/home/mjo/src/sage.git/local/var/lib/sage/scripts/sage_conf/spkg-requirements.txt'
> Warning: pip exited with status 0
> Removing stamp file '/home/mjo/src/sage.git/local/var/lib/sage/installed/sage_conf-10.2.rc4'
> ```

How do you provoke this error? I don't see it

**Comment by orlitzky** [comment]
> How do you provoke this error? I don't see it

```
$ ./bootstrap
$ ./configure <my long list of configure flags>
$ make sage_conf  # install python deps to prevent pip fails
$ sage -sh
$ sage-spkg sage_conf  # reinstall
$ sage-spkg-uninstall sage_conf
```

The reinstall may not be necessary, but those were the steps I used.

**Comment by orlitzky** [comment]
> ... so perhaps when using `--enable-system-site-packages`, you would want to disable build isolation?

I think so. Is there an easy way to do this across the board? If not, it's not that big of a deal. The `make` that most people use outputs the error in black and white and is a lot less scary.

**Comment by orlitzky** [comment]
My `./configure` flags can be found here:

https://gitweb.michael.orlitzky.com/?p=bash.d.git;a=blob;f=sage.sh;hb=HEAD

I've obviously got  `--enable-system-site-packages`, but maybe `--disable-editable` is relevant too, who knows.

**Comment by mkoeppe** [comment]
Oh, that's #36737. I'll merge it here

**Comment by mkoeppe** [comment]
> > ... so perhaps when using `--enable-system-site-packages`, you would want to disable build isolation?
> 
> I think so. Is there an easy way to do this across the board?

It all goes through `build/bin/sage-dist-helpers` (`sdh_pip_install`, `sdh_pip_editable_install`); for the first one, just set `build_isolation_option` to a different default for this case. (Separate PR)

**Comment by orlitzky** [comment]
Yep, that fixed it.

File: build/make/Makefile.in [hunk 1]
```diff
@@ -740,9 +740,9 @@ $(1)-$(4)-no-deps:
 		SAGE_SPKG_SCRIPTS=$$($(4))/var/lib/sage/scripts \
 		SAGE_INST_LOCAL=$$($(4)) \
 		SAGE_CHECK=$$(SAGE_CHECK_$(1)) \
-		sage-logger -p '$$(SAGE_ROOT)/build/pkgs/$(1)/spkg-install && if [ $$$$SAGE_CHECK != no -a -x $$(SAGE_ROOT)/build/pkgs/$(1)/spkg-check ]; then $$(SAGE_ROOT)/build/pkgs/$(1)/spkg-check; fi' '$$(SAGE_LOGS)/$(1)-$(2).log' && \
-	    rm -f "$$($(4))/$(SPKG_INST_RELDIR)/$(1)"-* && \
-	    touch "$$($(4))/$(SPKG_INST_RELDIR)/$(1)-$(2)"; \
+		sage-logger -p 'SAGE_CHECK=$$(SAGE_CHECK_$(1)) PATH=$$(SAGE_SRC)/bin:$$($(4))/bin:$$$$PATH $$(SAGE_SPKG) $$(SAGE_SPKG_OPTIONS) \
```
Comment (orlitzky) [inline build/make/Makefile.in:743]: > $$$$PATH :muscle:

**Review by orlitzky — APPROVED** [review]

**Comment by mkoeppe** [comment]
Thank you!

**Comment by vbraun** [comment]
merge conflict

**Pull request merged**